src/transport/manager/mod.rs

```diff
@@ -194,6 +194,44 @@ impl Stream for TransportContext {
     }
 }
 
+/// The IP dialing mode determines how the transport manager handles dialing
+/// IP addresses. It can either dial only global addresses or all addresses, including private IPs.
+#[derive(Debug, Default, Clone, Copy, Eq, PartialEq)]
+pub(crate) enum IpDialingMode {
+    /// Dial only global addresses.
+    GlobalOnly,
+
+    /// Dial all addresses, including private IPs.
+    ///
+    /// This setup is not recommended for production use-cases.
+    #[default]
+    All,
+}
+
+impl IpDialingMode {
+    /// Check if the IP address in the given `Multiaddr` is global.
+    pub fn is_address_global(address: &Multiaddr) -> bool {
+        let ip = match address.iter().next() {
+            Some(Protocol::Ip4(ip)) => ip_network::IpNetwork::from(ip),
+            Some(Protocol::Ip6(ip)) => ip_network::IpNetwork::from(ip),
+            Some(Protocol::Dns(_)) | Some(Protocol::Dns4(_)) | Some(Protocol::Dns6(_)) =>
+                return true,
```

Comment:
This will leak addresses like `/dns4/10.10.10.10/tcp/30333/...`, and also dns records can point to local IP addresses.